ENCD-6175-support-collection-series-and-transgenic-enhancer-experiments-in-cart

forresttanaka opened the pull request:
* I found a very small but non-zero window in which the user can add datasets to the cart before the cart initialized on page load, causing a JS crash. The two modifications to `initializeCart` prevent this by setting the cart to the in-progress state on page load, and initializing the `elements` property on savedCartObj — a frequent dependency for cart code. Normal cart startup operation includes turning off the in-progress state once the user can truly interact with the cart.
* Many changes in this ticket guard against future dataset types being added to the cart before someone fully integrates them into the cart.

~https://encd-6175-collection-series-cart-qa1-fytanaka.demo.encodedcc.org/~

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -422,11 +423,12 @@ const CartTools = ({
-        return types.includes(type) ? types : types.concat(type);
-    }, []);
+    // Make a list of all the allowed dataset types currently in the cart.
+    const usedDatasetTypes = new Set(
```
It might be a few nano seconds faster to do the filter first and then map. As-is, map operates on what is later filtered out.

forresttanaka replied:
I didn’t time a comparison, but I didn’t want to filter first because I’d have to search `allowedDatasetTypes` to find the matching type. By using map first, I can just do a direct lookup of `allowedDatasetTypes`.

zoldello reviewed (dismissed):
Code looks fine. I was able to add a collection cart on the UI. My comment is insignificant.

forresttanaka committed:
ENCD-6175-collection-series-cart

Individual series pages now immediately remove the series page as well as its related datasets.

forresttanaka committed:
ENCD-6175-collection-series-cart

Removing a series on search-result pages now removes the series and its related datasets.

forresttanaka committed:
ENCD-6175-collection-series-cart

Added a new tooltip for experiments in carts to indicate you can’t remove them except through the cart-view.

forresttanaka committed:
ENCD-6175-collection-series-cart

forresttanaka commented:
Updated for handling removing series and related datasets on individual pages, as described in Bonita’s QA.

Preview demo: https://encd-6175-collection-series-cart-pv2-fytanaka.demo.encodedcc.org/

zoldello reviewed (approved):
Code looks fine and has functionality to address issue Bonita brought up. I am curious, how do you rebase? I see like five commits rather than one.

forresttanaka commented:
I just rebased without squashing so it’s easier to tell which parts you’ve reviewed and which you haven’t.

zoldello reviewed (approved):
Let's have Bonita give this a go. Code looks fine.
